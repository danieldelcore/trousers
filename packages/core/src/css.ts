import hash from '@trousers/hash';

import { Definition, CSSObject } from './types';
import themify from './themify';

export interface CollectorReturn {
    _get: () => Definition[];
    modifier: (id: string, styles: CSSObject) => CollectorReturn;
    global: (styles: CSSObject) => CollectorReturn;
    theme: (theme: Record<string, any>) => CollectorReturn;
}

function css(styles: CSSObject): CollectorReturn;
function css(
    idOrStyle: string | CSSObject,
    styles?: CSSObject,
): CollectorReturn;
function css(
    idOrStyle: string | CSSObject,
    styles?: CSSObject,
): CollectorReturn {
    let id = typeof idOrStyle === 'string' ? idOrStyle : '';
    let styleObject = (typeof idOrStyle === 'string'
        ? styles
        : idOrStyle) as CSSObject;

    const elementId = `${id}-${hash(JSON.stringify(styleObject))}`;
    const styleMap: Definition[] = [
        {
            id,
            type: 'element',
            className: `.${elementId}`,
            styles: styleObject,
        },
    ];

    const self: CollectorReturn = {
        _get: () => styleMap,
        modifier: (modifierId, modifierStyles) => {
            styleMap.push({
                id: modifierId,
                type: 'modifier',
                className: `.${elementId}--${modifierId}-${hash(
                    JSON.stringify(modifierStyles),
                )}`,
                styles: modifierStyles,
            });

            return self;
        },
        global: globalStyles => {
            styleMap.push({
                id: `global-${id}-${hash(JSON.stringify(globalStyles))}`,
                type: 'global',
                className: '',
                styles: globalStyles,
            });

            return self;
        },
        theme: theme => {
            const themeHash = hash(JSON.stringify(theme));
            styleMap.push({
                id: `theme-${id}-${themeHash}`,
                type: 'theme',
                className: `.theme-${id}-${themeHash}`,
                styles: themify(theme),
            });

            return self;
        },
    };

    return self;
}

export type Collector = typeof css;

export default css;
