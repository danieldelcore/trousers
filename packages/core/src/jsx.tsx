import { createElement, hasOwnProperty, ElementType, ReactNode } from 'react';

import { StyleCollector, CSSProps } from '@trousers/utils';
import useStyles from './useStyles';

import './types';

const jsx = <
    Props extends { css: StyleCollector<Theme> | CSSProps },
    Theme extends {} = {}
>(
    type: ElementType<Omit<Props, 'css'>>,
    props: Props,
    ...children: ReactNode[]
) => {
    if (props == null || !hasOwnProperty.call(props, 'css')) {
        return createElement(type, props, ...children);
    }

    const { css, ...rest } = props;

    // eslint-disable-next-line react-hooks/rules-of-hooks
    const className = useStyles<Theme>(props.css);

    return createElement(type, { ...rest, className }, ...children);
};

export default jsx;
export { CSSProps };
