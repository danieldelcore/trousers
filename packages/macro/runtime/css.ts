interface Definition {
    type: 'element' | 'modifier' | 'global' | 'theme';
    id: string;
    styles: string;
}

interface CollectorReturn {
    modifier: (id: string, styles: string) => CollectorReturn;
    global: (id: string, styles: string) => CollectorReturn;
    theme: (id: string, styles: string) => CollectorReturn;
    _get: () => Definition[];
}

function css(elementId: string, elementStyles: string) {
    const styleMap: Definition[] = [
        {
            type: 'element',
            id: elementId,
            styles: elementStyles,
        },
    ];

    const self: CollectorReturn = {
        _get: () => styleMap,
        modifier: (modifierId, modifierStyles) => {
            styleMap.push({
                type: 'modifier',
                id: modifierId,
                styles: modifierStyles,
            });

            return self;
        },
        global: (globalId, globalStyles) => {
            styleMap.push({
                type: 'global',
                id: globalId,
                styles: globalStyles,
            });

            return self;
        },
        theme: (themeId, themeStyles) => {
            styleMap.push({
                type: 'theme',
                id: themeId,
                styles: themeStyles,
            });

            return self;
        },
    };

    return self;
}

export default css;
