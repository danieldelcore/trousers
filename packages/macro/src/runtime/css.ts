interface Definition {
    type: 'element' | 'modifier' | 'global' | 'theme';
    id: string;
    styles: Record<string, string>;
}

export interface CollectorReturn {
    modifier: (
        id: Definition['id'],
        styles: Definition['styles'],
    ) => CollectorReturn;
    global: (
        id: Definition['id'],
        styles: Definition['styles'],
    ) => CollectorReturn;
    theme: (
        id: Definition['id'],
        styles: Definition['styles'],
    ) => CollectorReturn;
    _get: () => Definition[];
}

function css(elementId: Definition['id'], elementStyles: Definition['styles']) {
    const styleMap: Definition[] = [
        {
            type: 'element',
            id: elementId,
            styles: elementStyles,
        },
    ];

    const self: CollectorReturn = {
        _get: () => styleMap,
        modifier: (modifierId, modifierStyles) => {
            styleMap.push({
                type: 'modifier',
                id: modifierId,
                styles: modifierStyles,
            });

            return self;
        },
        global: (globalId, globalStyles) => {
            styleMap.push({
                type: 'global',
                id: globalId,
                styles: globalStyles,
            });

            return self;
        },
        theme: (themeId, themeStyles) => {
            styleMap.push({
                type: 'theme',
                id: themeId,
                styles: themeStyles,
            });

            return self;
        },
    };

    return self;
}

export default css;
