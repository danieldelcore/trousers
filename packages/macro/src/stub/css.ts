/* eslint-disable @typescript-eslint/no-unused-vars */
import {
    CSSObject,
    Collector as CoreCollector,
    CollectorReturn as CoreCollectorReturn,
} from '@trousers/core';

export type CollectorReturn = CoreCollectorReturn;
export type Collector = CoreCollector;

function css(_styles: CSSObject): CollectorReturn;
function css(
    _idOrStyle: string | CSSObject,
    _styles?: CSSObject,
): CollectorReturn;
function css(
    _idOrStyle: string | CSSObject,
    _styles?: CSSObject,
): CollectorReturn {
    const self: CollectorReturn = {
        modifier: (_modifierId, _modifierStyles) => self,
        global: _globalStyles => self,
        theme: _theme => self,
        _get: () => [],
    };

    return self;
}

export default css;
