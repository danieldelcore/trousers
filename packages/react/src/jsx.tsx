import './global-types';
import React, {
    createElement,
    hasOwnProperty,
    ElementType,
    ReactNode,
    useLayoutEffect,
    Fragment,
} from 'react';

import { TrousersProps, process, isBrowser } from '@trousers/core';
import sheet from '@trousers/sheet';

let styleSheet: ReturnType<typeof sheet> | null = null;
if (isBrowser()) {
    const headElement = document.getElementsByTagName('head')[0];
    styleSheet = sheet(headElement, 'data-trousers');
}

const jsx = <Props extends TrousersProps>(
    type: ElementType<Omit<Props, 'css'>>,
    props: Props,
    ...children: ReactNode[]
) => {
    if (props == null || !hasOwnProperty.call(props, 'css'))
        return createElement(type, props, ...children);

    //TODO: might be good to memo here
    const definitions = props
        .css!._get()
        .filter(
            ({ id, type }) =>
                type !== 'modifier' ||
                (props.hasOwnProperty(`$${id}`) && !!props[`$${id}`]),
        );

    const classes = definitions
        .map(({ className }) => className.substring(1))
        .join(' ')
        .trim();

    const cleanProps = Object.keys(props)
        .filter(key => !key.startsWith('$') && key !== 'css')
        .reduce((obj: Record<string, any>, key: string) => {
            obj[key] = props[key];
            return obj;
        }, {});

    const Element = createElement(
        type,
        // @ts-ignore
        {
            ...cleanProps,
            className: classes,
        },
        ...children,
    );

    if (!isBrowser()) {
        const styles = definitions
            .map(({ className, styles }) =>
                Object.entries(process(className, styles)).reduce(
                    (accum, [key, value]) => `${accum}${key} {${value}}`,
                    '',
                ),
            )
            .join('\n');

        return (
            <Fragment>
                <style dangerouslySetInnerHTML={{ __html: styles }} />
                {Element}
            </Fragment>
        );
    }

    // eslint-disable-next-line react-hooks/rules-of-hooks
    useLayoutEffect(() => {
        const cleanUp: string[] = [];

        definitions
            .filter(({ className }) => styleSheet && !styleSheet.has(className)) // this doesn't stop all selectors
            .forEach(({ id, className, styles, type }) =>
                Object.entries(process(className, styles)).forEach(
                    ([key, value]) => {
                        if (type === 'global') {
                            cleanUp.push(id + key);
                        }

                        return (
                            styleSheet &&
                            styleSheet.mount(
                                id + key,
                                `${key}{${value}}`,
                                type === 'global',
                            )
                        );
                    },
                ),
            );

        return () => {
            cleanUp.forEach(id => styleSheet && styleSheet.unmount(id));
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [classes]);

    return Element;
};

export default jsx;
