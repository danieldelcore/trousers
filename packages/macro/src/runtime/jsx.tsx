import React, { FC, createElement, useLayoutEffect, Fragment } from 'react';

import { isBrowser } from '@trousers/core';
import Sheet from '@trousers/sheet';

import { CollectorReturn } from './css';

export interface TrousersProps {
    css?: CollectorReturn;
    [key: string]: any; // Will need to use template literal types in ts 4.1
}

let sheet: ReturnType<typeof Sheet> | null = null;
if (isBrowser()) {
    const headElement = document.getElementsByTagName('head')[0];
    sheet = Sheet(headElement, 'data-trousers');
}

const TrousersNested: FC<TrousersProps> = props => {
    const definitions = props
        .css!._get()
        .filter(
            ({ id, type }) =>
                type !== 'modifier' ||
                (props.hasOwnProperty(`$${id}`) && !!props[`$${id}`]),
        );

    const classes = definitions
        .map(({ styles }) => Object.keys(styles)[0].substring(1))
        .join(' ')
        .trim();

    const cleanProps = Object.keys(props)
        .filter(
            key =>
                !key.startsWith('$') &&
                key !== 'css' &&
                key !== 'elementType' &&
                key !== 'children',
        )
        .reduce((obj: Record<string, any>, key: string) => {
            obj[key] = props[key];
            return obj;
        }, {});

    const Element = createElement(
        props.elementType,
        {
            ...cleanProps,
            className: classes,
        },
        props.children,
    );

    if (!isBrowser()) {
        const styles = definitions
            .map(({ styles }) =>
                Object.entries(styles).reduce(
                    (accum, [key, value]) => `${accum}${key} {${value}}`,
                    '',
                ),
            )
            .join('\n');

        return (
            <Fragment>
                <style dangerouslySetInnerHTML={{ __html: styles }} />
                {Element}
            </Fragment>
        );
    }

    // eslint-disable-next-line react-hooks/rules-of-hooks
    useLayoutEffect(() => {
        const cleanUp: string[] = [];

        definitions
            .filter(({ styles }) => sheet && !sheet.has(Object.keys(styles)[0])) // this doesn't stop all selectors
            .forEach(({ id, styles, type }) =>
                Object.entries(styles).forEach(([key, value]) => {
                    if (type === 'global') {
                        cleanUp.push(id + key);
                    }

                    return (
                        sheet &&
                        sheet.mount(
                            id + key,
                            `${key}{${value}}`,
                            type === 'global',
                        )
                    );
                }),
            );

        return () => {
            cleanUp.forEach(id => sheet && sheet.unmount(id));
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [classes]);

    return Element;
};

export default TrousersNested;
