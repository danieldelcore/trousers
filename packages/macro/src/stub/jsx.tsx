import '../runtime/global-types';
import { createElement, ElementType, ReactNode } from 'react';

import { TrousersProps } from '@trousers/core';

const jsx = <Props extends TrousersProps>(
    type: ElementType<Omit<Props, 'css'>>,
    props: Props,
    ...children: ReactNode[]
) => createElement(type, props, ...children);

export default jsx;
