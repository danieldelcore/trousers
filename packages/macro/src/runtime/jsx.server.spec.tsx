/** @jsx jsx */
import css from './css';
import jsx from './jsx';
import { render } from '@testing-library/react';

jest.mock('@trousers/core', () => ({
    __esModule: true,
    ...(jest.requireActual('@trousers/core') as object),
    isBrowser: () => false,
}));

describe('jsx (ssr mode)', () => {
    it('renders element styles', () => {
        const styles = css('element', { '.element-2313942302': 'color: red;' });
        const { container, getByTestId } = render(
            <div css={styles} data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"element-2313942302"`,
        );
        expect(container).toMatchSnapshot();
    });

    it('renders element styles without id', () => {
        const styles = css('', { '.2313942302': 'color: red;' });
        const { container, getByTestId } = render(
            <div css={styles} data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"2313942302"`,
        );
        expect(container).toMatchSnapshot();
    });

    it('does not render modifier if prop is not provided', () => {
        const styles = css('element', {
            '.element-2313942302': 'color: red;',
        }).modifier('primary', { '.primary-2561700995': 'color: blue;' });

        const { container, getByTestId } = render(
            <div css={styles} data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"element-2313942302"`,
        );
        expect(container).toMatchSnapshot();
    });

    it('does not render modifier if prop false', () => {
        const styles = css('element', {
            '.element-2313942302': 'color: red;',
        }).modifier('primary', {
            '.primary-2561700995': 'color: blue;',
        });

        const { container, getByTestId } = render(
            <div css={styles} $primary={false} data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"element-2313942302"`,
        );
        expect(container).toMatchSnapshot();
    });

    it('does not render modifier if prop undefined', () => {
        const styles = css('element', {
            '.element-2313942302': 'color: red;',
        }).modifier('primary', {
            '.primary-2561700995': 'color: blue;',
        });

        const { container, getByTestId } = render(
            <div css={styles} $primary={undefined} data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"element-2313942302"`,
        );
        expect(container).toMatchSnapshot();
    });

    it('renders element and modifier styles', () => {
        const styles = css('element', {
            '.element-2313942302': 'color: red;',
        }).modifier('primary', {
            '.primary-2561700995': 'color: blue;',
        });

        const { container, getByTestId } = render(
            <div css={styles} $primary data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"element-2313942302 primary-2561700995"`,
        );
        expect(container).toMatchSnapshot();
    });

    it('renders element and modifier styles without id', () => {
        const styles = css('element', {
            '.2313942302': 'color: red;',
        }).modifier('primary', {
            '.primary-2561700995': 'color: blue;',
        });
        const { container, getByTestId } = render(
            <div css={styles} $primary data-testid="test" />,
        );
        expect(getByTestId('test').className).toMatchInlineSnapshot(
            `"2313942302 primary-2561700995"`,
        );
        expect(container).toMatchSnapshot();
    });
});
