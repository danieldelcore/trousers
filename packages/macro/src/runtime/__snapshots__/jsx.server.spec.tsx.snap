// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsx (ssr mode) does not render modifier if prop false 2`] = `
<div>
  <style>
    .element-2313942302 {color: red;}
  </style>
  <div
    class="element-2313942302"
    data-testid="test"
  />
</div>
`;

exports[`jsx (ssr mode) does not render modifier if prop is not provided 2`] = `
<div>
  <style>
    .element-2313942302 {color: red;}
  </style>
  <div
    class="element-2313942302"
    data-testid="test"
  />
</div>
`;

exports[`jsx (ssr mode) does not render modifier if prop undefined 2`] = `
<div>
  <style>
    .element-2313942302 {color: red;}
  </style>
  <div
    class="element-2313942302"
    data-testid="test"
  />
</div>
`;

exports[`jsx (ssr mode) renders element and modifier styles 2`] = `
<div>
  <style>
    .element-2313942302 {color: red;}
.primary-2561700995 {color: blue;}
  </style>
  <div
    class="element-2313942302 primary-2561700995"
    data-testid="test"
  />
</div>
`;

exports[`jsx (ssr mode) renders element and modifier styles without id 2`] = `
<div>
  <style>
    .2313942302 {color: red;}
.primary-2561700995 {color: blue;}
  </style>
  <div
    class="2313942302 primary-2561700995"
    data-testid="test"
  />
</div>
`;

exports[`jsx (ssr mode) renders element styles 2`] = `
<div>
  <style>
    .element-2313942302 {color: red;}
  </style>
  <div
    class="element-2313942302"
    data-testid="test"
  />
</div>
`;

exports[`jsx (ssr mode) renders element styles without id 2`] = `
<div>
  <style>
    .2313942302 {color: red;}
  </style>
  <div
    class="2313942302"
    data-testid="test"
  />
</div>
`;
